### PR TITLE
[MRG] Fix plotting of annotations.

### PR DESCRIPTION
**Pull request by jaeilepp**
Closes https://github.com/mne-tools/mne-python/issues/3509.
I believe there was a problem with syncing the annotations when cropping and concatenating data with ``first_samp`` present. I'll still have to make some checks before merging.

### CODE REVIEW

File: mne/io/fiff/tests/test_raw_fiff.py [hunk 1]
```diff
@@ -1179,6 +1179,17 @@ def test_save():
...
+
+    # test that annotations are in sync after cropping and concatenating
+    annot = Annotations([2.5], [2], ['test'])
+    r1 = raw.copy().crop(2.5, 7.5)
+    r1.annotations = annot
+    r2 = raw.copy().crop(12.5, 17.5)
+    r2.annotations = annot
+    raw = concatenate_raws([r1, r2])
+    onsets = raw.annotations.onset
+    assert_array_almost_equal([2.5, 7.5], onsets, decimal=2)
```
Comment (agramfort) [inline mne/io/fiff/tests/test_raw_fiff.py]: can you document here why 7.5 is the expected result?

Comment (mmagnuski) [inline mne/io/fiff/tests/test_raw_fiff.py]: first raw segment has 5 s and the annotation in second time segments starts at 2.5 - so when concatenated the second annotation should start at 5 + 2.5

Comment (mmagnuski) [inline mne/io/fiff/tests/test_raw_fiff.py]: but I agree there should be a comment about this in the code.

**Comment by jaeilepp** [comment]
I'm thinking that maybe we should prevent annotating segments outside the area of the data (it currently isn't). Now it is possible that this creates annotations in wrong positions when cropping and concatenating the data. This would mean that the data has to be concatenated before annotating split data, but I think that's okay.

**Comment by agramfort** [comment]
+1

**Comment by jaeilepp** [comment]
I think this finally works as it's supposed to. I had to make quite a bit of changes as it previously did not check whether the annotations were in the data range causing the concatenation to effectively break.

File: mne/annotations.py [hunk 2]
```diff
@@ -101,11 +101,10 @@ def _combine_annotations(annotations, last_samps, first_samps, sfreq):
         old_description = annotations[0].description
         old_orig_time = annotations[0].orig_time
 
-    if annotations[1].orig_time is None:
-        onset = (annotations[1].onset +
-                 (sum(last_samps[:-1]) - sum(first_samps[:-1])) / sfreq)
-    else:
-        onset = annotations[1].onset
+    extra_samps = len(first_samps) - 1  # Account for sample 0
```
Comment (jaeilepp) [inline mne/annotations.py]: Previously the annotations were shifting by 1 sample when concatenated. This fixes it.

File: doc/whats_new.rst [hunk 3]
```diff
@@ -110,6 +110,8 @@ BUG
 
         - Fix comparisons of filter settings for determining "strictest" filter
 
+    - Fixes to plotting and timing of :class:`mne.Annotations <annotations>` and restricted addition of annotations outside data range to prevent problems with cropping and concatenating data by `Jaakko Leppakangas`_
```
Comment (Eric89GXL) [inline doc/whats_new.rst]: "fixed" to match all the other tenses here

Comment (Eric89GXL) [inline doc/whats_new.rst]: And check the rendering, I think you mean:
```
:class:`annotation <mne.Annotations>`
```
but really this is good enough:
```
:class:`Annotations`
```
We're already in the `mne` namespace in this file

File: mne/annotations.py [hunk 4]
```diff
@@ -120,13 +119,14 @@ def _onset_to_seconds(raw, onset):
         meas_date = 0
     elif not np.isscalar(meas_date):
         if len(meas_date) > 1:
-            meas_date[1] /= 1e6
-        meas_date = meas_date.sum()
+            meas_date = float(meas_date[0]) + meas_date[1] / 1000000.
```
Comment (Eric89GXL) [inline mne/annotations.py]: you shouldn't need this `float` cast it will happen automatically because the right operand of `+` is a float

**Comment by Eric89GXL** [comment]
Otherwise LGTM, Travis errors seem unrelated

File: mne/io/base.py [hunk 5]
```diff
@@ -585,6 +585,51 @@ def last_samp(self):
     def _raw_lengths(self):
         return [l - f + 1 for f, l in zip(self._first_samps, self._last_samps)]
 
+    @property
+    def annotations(self):
```
Comment (wmvanvliet) [inline mne/io/base.py]: doc?

File: mne/io/base.py [hunk 6]
```diff
@@ -585,6 +585,51 @@ def last_samp(self):
...
+        return self._annotations
+
+    @annotations.setter
+    def annotations(self, value):
+        """Setter for annotations. Checks if they are inside the data range."""
```
Comment (wmvanvliet) [inline mne/io/base.py]: explanation of parameters?

**Comment by agramfort** [comment]
can you rebase to restart CIs?

**Comment by jaeilepp** [comment]
All green now.

**Comment by agramfort** [comment]
thx @jaeilepp

**Pull request merged**